### PR TITLE
modify OAP-accumulators to SQL-Level(sub-PR of #506)

### PR DESCRIPTION
**Pull request by TigerSong**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

* modify OAP-accumulators to SQL-Level
* add more accumulators
   * Task-level
   * Row-level

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

UT
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala [hunk 1]
```diff
@@ -200,8 +201,16 @@ private[oap] class OapDataReader(
...
+    def getFullScanIter: OapIterator[InternalRow] = {
+      val start = System.currentTimeMillis()
+      val iter = fileScanner.iterator(conf, requiredIds)
+      val end = System.currentTimeMillis()
+      logDebug("Contruct File Iterator: " + (end - start) + " ms")
+      iter
+    }
+
     filterScanners match {
-      case Some(indexScanners) if indexScanners.indexIsAvailable(path, conf) =>
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: keep this guard and then case other Some(indexScanners) => getFullScanIter ?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala [hunk 2]
```diff
@@ -200,8 +201,16 @@ private[oap] class OapDataReader(
     // TODO how to save the additional FS operation to get the Split size
     val fileScanner = DataFile(path.toString, meta.schema, meta.dataReaderClassName, conf)
 
+    def getFullScanIter: OapIterator[InternalRow] = {
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: Is this method named fullScanIter better?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala [hunk 3]
```diff
@@ -200,8 +201,16 @@ private[oap] class OapDataReader(
...
+      val start = System.currentTimeMillis()
+      val iter = fileScanner.iterator(conf, requiredIds)
+      val end = System.currentTimeMillis()
+      logDebug("Contruct File Iterator: " + (end - start) + " ms")
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: It's debug log, should we call System.currentTimeMillis() only debug.isEnable == true?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala [hunk 4]
```diff
@@ -191,7 +191,8 @@ private[oap] class OapDataReader(
   filterScanners: Option[IndexScanners],
   requiredIds: Array[Int]) extends Logging {
 
-  var selectedRows: Option[Long] = None
+  var rowsReadWhenHitIndex: Option[Long] = None
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: really need to be packed by Option? when hitIndex the values is rows.length,  other cases use default 0?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/OapIndexInfoStatusSerDe.scala [hunk 5]
```diff
@@ -41,7 +41,7 @@ private[oap] object OapIndexInfoStatusSerDe extends SerDe[String, Seq[OapIndexIn
 
   override def deserialize(json: String): Seq[OapIndexInfoStatus] = {
     (parse(json) \ "oapIndexInfoStatusRawDataArray").extract[List[JValue]].map(
-      indexStatusRawDataFromJson)
+      indexStatusRawDataFromJson _)
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/OapIndexInfoStatusSerDe.scala]: Why add this "_"?

**Comment by LuciferYang** [comment]
Which UT can test this issue ? List new Suite name or Suite Class name?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala (hunk 4 above)
Comment (TigerSong) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: 0 is a available value. Here 'None' means it miss/ignore index. I think None or -1 is ok, and i choose None

Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: got it

**Comment by LuciferYang** [comment]
LGTM~

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala [hunk 6]
```diff
@@ -204,8 +204,7 @@ private[oap] class OapDataReader(
     def fullScan: OapIterator[InternalRow] = {
       val start = if (log.isDebugEnabled) System.currentTimeMillis() else 0
       val iter = fileScanner.iterator(conf, requiredIds)
-      val end = if (log.isDebugEnabled) System.currentTimeMillis() else 0
-      logDebug("Contruct File Iterator: " + (end - start) + " ms")
+      logDebug("Construct File Iterator: " + (System.currentTimeMillis() - start) + " ms")
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: @TigerSong  There is a problem, System.currentTimeMillis() will be calculated at msg func run time but not actually "fileScanner.iterator" end time......

File: src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 7]
```diff
@@ -249,7 +250,31 @@ case class FileSourceScanExec(
...
+    format.rowsSkippedWhenHitIndex = Some(metrics("rowsSkippedWhenHitIndex"))
+    format.rowsReadWhenIgnoreIndex = Some(metrics("rowsReadWhenIgnoreIndex"))
+    format.rowsReadWhenMissIndex = Some(metrics("rowsReadWhenMissIndex"))
+  }
+
   private lazy val inputRDD: RDD[InternalRow] = {
+    // init accumulator before buildReader
+    relation.fileFormat match {
+      case format: OapFileFormat =>
+        initOapAccumulator(format)
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: can we move this function to OapFileFormat?
I mean, change`initOapAccumulator(format)` to `format.initOapAccumulator`

File: src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 8]
```diff
@@ -23,16 +23,17 @@ import org.apache.commons.lang3.StringUtils
 import org.apache.hadoop.fs.{BlockLocation, FileStatus, LocatedFileStatus, Path}
 
 import org.apache.spark.rdd.RDD
-import org.apache.spark.sql.{AnalysisException, SparkSession}
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Please pay attention to Idea's import optimization.
Don't remove this AnalysisException

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 9]
```diff
@@ -39,18 +39,41 @@ import org.apache.spark.sql.execution.datasources.oap.filecache.DataFileHandleCa
...
+  var hitIndexTasks: Option[SQLMetric] = None
+  var ignoreIndexTasks: Option[SQLMetric] = None
+  var missIndexTasks: Option[SQLMetric] = None
+
+  var totalRows: Option[SQLMetric] = None
+  var rowsSkippedForStatistic: Option[SQLMetric] = None
+  var rowsReadWhenHitIndex: Option[SQLMetric] = None
+  var rowsSkippedWhenHitIndex: Option[SQLMetric] = None
+  var rowsReadWhenIgnoreIndex: Option[SQLMetric] = None
+  var rowsReadWhenMissIndex: Option[SQLMetric] = None
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: all above members should be wrap into a class.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 10]
```diff
@@ -287,36 +307,33 @@ private[sql] class OapFileFormat extends FileFormat
...
-            selectedRows.add(reader.selectedRows.getOrElse(totalRows))
-            skippedRows.add(totalRows - reader.selectedRows.getOrElse(totalRows))
-
-            val fullSchema = requiredSchema.toAttributes ++ partitionSchema.toAttributes
-            val joinedRow = new JoinedRow()
-            val appendPartitionColumns = GenerateUnsafeProjection.generate(fullSchema, fullSchema)
-
-            iter.map(d => appendPartitionColumns(joinedRow(d, file.partitionValues)))
+          totalRows.foreach(_.add(rowsInTotal))
+          totalTasks.foreach(_.add(1L))
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: as I mentioned above, all members should be wrapped into a class.
and the members value modification should via a function.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 11]
```diff
@@ -326,6 +343,27 @@ private[sql] class OapFileFormat extends FileFormat
     }
   }
 
+  private def metrics(r: Option[OapDataReader], rowsInTotal: Long): Unit = {
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: please leave some descriptions about this function. meaning of parameters and job this this function.
also, give a description of the pattern matching. seems a lot of condition branches here.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala (hunk 4 above)
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: In my understanding, this member should be modified the the owner itself. and can only be read by other classes.
So, can you change it to private? and create a `getXX` function.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala [hunk 12]
```diff
@@ -200,6 +201,14 @@ private[oap] class OapDataReader(
     // TODO how to save the additional FS operation to get the Split size
     val fileScanner = DataFile(path.toString, meta.schema, meta.dataReaderClassName, conf)
 
+    def fullScan: OapIterator[InternalRow] = {
+      val start = if (log.isDebugEnabled) System.currentTimeMillis else 0
+      val iter = fileScanner.iterator(conf, requiredIds)
+      val end = if (log.isDebugEnabled) System.currentTimeMillis else 0
+      logDebug("Construct File Iterator: " + (end - start) + " ms")
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: please help change this by using `Utils.getUsedTimeMs()`

File: src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 7 above)
Comment (TigerSong) [reply src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: there will be a little problem. When assign reference to format.totalRows, we need assign the name, like 'totalRows', 'rowsSkippedForStatistic', etc.
if we change, we need keep the name same in two different file.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala [hunk 13]
```diff
@@ -191,7 +191,11 @@ private[oap] class OapDataReader(
   filterScanners: Option[IndexScanners],
   requiredIds: Array[Int]) extends Logging {
 
-  var selectedRows: Option[Long] = None
+  private var _rowsReadWhenHitIndex: Option[Long] = None
+  private var _ignoreIndex: Boolean = false
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: Change these to:
```
private var rowRead: Long = 0
private var indexState: INDEX_STATE = NO_INDEX
// INDEX_STAT = {NO_INDEX, HIT_INDEX, IGNORE_INDEX} 
```
Your current implementation, each parameter has multiple meaning:
rowsReadWhenHitIndex = None => index miss or index ignore
rowsReadWhenHitIndex = Some(x) => index hit, row count
_ignoreIndex = true => index ignore
_ignoreIndex = false => index hit, or index miss

This causes you code logic in `metrics` too complex.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 14]
```diff
@@ -326,6 +404,27 @@ private[sql] class OapFileFormat extends FileFormat
...
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: change this to `updateIndexAndRowRead()` and move to metric class

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 15]
```diff
@@ -39,18 +39,100 @@ import org.apache.spark.sql.execution.datasources.oap.filecache.DataFileHandleCa
...
+      // set row-level Accumulator
+      _totalRows = Some(metrics("totalRows"))
+      _rowsSkippedForStatistic = Some(metrics("rowsSkippedForStatistic"))
+      _rowsReadWhenHitIndex = Some(metrics("rowsReadWhenHitIndex"))
+      _rowsSkippedWhenHitIndex = Some(metrics("rowsSkippedWhenHitIndex"))
+      _rowsReadWhenIgnoreIndex = Some(metrics("rowsReadWhenIgnoreIndex"))
+      _rowsReadWhenMissIndex = Some(metrics("rowsReadWhenMissIndex"))
+    }
+
+    def totalRows(rows: Long): Unit = {
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: change to `setTotalRows` or `updateTotalRows`

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 16]
```diff
@@ -524,4 +623,30 @@ private[sql] object OapFileFormat {
       OAP_INDEX_SCAN_NUM_OPTION_KEY ::
       OAP_INDEX_GROUP_BY_OPTION_KEY :: Nil
   }
+
+  def oapMetrics(sparkContext: SparkContext): Map[String, SQLMetric] =
+    Map("totalTasks" ->
+      SQLMetrics.createMetric(sparkContext, "OAP:tasks in total"),
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: OAP: total tasks

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 17]
```diff
@@ -524,4 +623,30 @@ private[sql] object OapFileFormat {
...
+      "skipForStatisticTasks" ->
+        SQLMetrics.createMetric(sparkContext, "OAP:tasks can skip"),
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: OAP: tasks skipped by statistics

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 18]
```diff
@@ -524,4 +623,30 @@ private[sql] object OapFileFormat {
...
+      "hitIndexTasks" ->
+        SQLMetrics.createMetric(sparkContext, "OAP:tasks hit index"),
+      "ignoreIndexTasks" ->
+        SQLMetrics.createMetric(sparkContext, "OAP:tasks hit-ignore index"),
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: OAP: tasks hit but ignore index
or just
OAP: task ignore index

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 19]
```diff
@@ -524,4 +623,30 @@ private[sql] object OapFileFormat {
...
+      "missIndexTasks" ->
+        SQLMetrics.createMetric(sparkContext, "OAP:tasks miss index"),
+
+      "totalRows" ->
+        SQLMetrics.createMetric(sparkContext, "OAP:rows in total"),
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: OAP: total rows

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 20]
```diff
@@ -524,4 +623,30 @@ private[sql] object OapFileFormat {
...
+      "rowsSkippedForStatistic" ->
+        SQLMetrics.createMetric(sparkContext, "OAP:rows in skip-tasks"),
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: OAP: rows skipped when task skipped

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 21]
```diff
@@ -39,18 +39,100 @@ import org.apache.spark.sql.execution.datasources.oap.filecache.DataFileHandleCa
...
+
+    def totalRows: Option[SQLMetric] = _totalRows
+    def rowsSkippedForStatistic: Option[SQLMetric] = _rowsSkippedForStatistic
+    def rowsReadWhenHitIndex: Option[SQLMetric] = _rowsReadWhenHitIndex
+    def rowsSkippedWhenHitIndex: Option[SQLMetric] = _rowsSkippedWhenHitIndex
+    def rowsReadWhenIgnoreIndex: Option[SQLMetric] = _rowsReadWhenIgnoreIndex
+    def rowsReadWhenMissIndex: Option[SQLMetric] = _rowsReadWhenMissIndex
+
+    def initMetrics(metrics: Map[String, SQLMetric]): Unit = {
+      _totalTasks = Some(metrics("totalTasks"))
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: Please add description for each metric.

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala [hunk 22]
```diff
@@ -934,8 +935,53 @@ class FilterSuite extends QueryTest with SharedOapContext with BeforeAndAfterEac
     sql("drop oindex idx2 on parquet_test")
   }
 
-  test("skipped rows and selected rows") {
-    val rowRDD = spark.sparkContext.parallelize(1 to 100, 3).map(i =>
+
+  def getValue(metrics: Option[SQLMetric]): Long = metrics.map(_.value).getOrElse(0L)
```
Comment (linhong-intel) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala]: Please create your own test suite class.
You can use `OapMetricsSuite`

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala (hunk 21 above)
Comment (TigerSong) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: described in oapMetrics()

**Comment by linhong-intel** [comment]
merge this as a requested feature by Baidu.
We'll come up with a formal metrics framework after this PR.

**Pull request merged**